runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -76,6 +76,19 @@
 #define AIX_NULL_ZERO_AREA_SIZE 256
 #define NUM_PICS 3
 
+#define LOCK_INC_DEC_VALUE                                OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT
+#define LOCK_RESERVATION_BIT                              OBJECT_HEADER_LOCK_RESERVED
+
+#define LOCK_RES_PRIMITIVE_ENTER_MASK                     (OBJECT_HEADER_LOCK_RECURSION_MASK | OBJECT_HEADER_LOCK_FLC)
+#define LOCK_RES_PRIMITIVE_EXIT_MASK                      (OBJECT_HEADER_LOCK_BITS_MASK & ~OBJECT_HEADER_LOCK_RECURSION_MASK)
+#define LOCK_RES_NON_PRIMITIVE_ENTER_MASK                 (OBJECT_HEADER_LOCK_RECURSION_MASK & ~OBJECT_HEADER_LOCK_LAST_RECURSION_BIT)
+#define LOCK_RES_NON_PRIMITIVE_EXIT_MASK                  (OBJECT_HEADER_LOCK_RECURSION_MASK & ~OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT)
+
+#define LOCK_RES_OWNING_COMPLEMENT                        (OBJECT_HEADER_LOCK_RECURSION_MASK | OBJECT_HEADER_LOCK_FLC)
+
+#define LOCK_NON_PRIMITIVE_ENTER_IGNORE_MASK              ((OBJECT_HEADER_LOCK_RECURSION_MASK & ~OBJECT_HEADER_LOCK_LAST_RECURSION_BIT) | OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_FLC)
+#define LOCK_NON_PRIMITIVE_EXIT_IGNORE_MASK               (OBJECT_HEADER_LOCK_RECURSION_MASK | OBJECT_HEADER_LOCK_RESERVED | OBJECT_HEADER_LOCK_FLC)
```

Comment:
Masks used in rotate-masking instructions have to be a single-range circularly.  I assumed you already made sure of this.

Reply:
Most of the masks only deal with the lowest 8 bits. The exceptions are the ones that need to look at the TID stored in the lockword. I've made sure that the masks used by instructions like `rlwinm` are compatible with the instruction.

Reply:
I double checked this and I actually made a mistake earlier. Those masks will not work with `rlwinm` due to the position of the Learning bit. So I will have to remove that path.